julia_con/RelatedCon/src/RelatedCon.jl

```diff
@@ -40,43 +36,52 @@ function maxindex!(xs::Vector, maxs)
             top = maxs[1][2]
         end
     end
-    maxs
+    return
 end
 
-function related!(posts, tagmap, relatedposts)
+function related(posts)
+    Ts = (UInt8, UInt16, UInt32, UInt64)
+    i = findfirst(T -> length(posts) < typemax(T), Ts)
+    return related(Ts[i], posts)
```

Comment:
Is this really a good idea? I haven't checked but I'd worry that the compiler wouldn't like this from a type stability POV. The old way this was done was to do
```
for T in (UInt8, UInt16, UInt32, UInt64)
    if length(posts) < typemax(T)
        related(T, posts)
    end
end
```
which should be a little easier for the compiler to reason about

Reply:
let's use this way then in both versions. Let's see risk averse :D

Reply:
```julia
julia> function f_long(posts)
           for T in (UInt8, UInt16, UInt32, UInt64)
               if length(posts) < typemax(T)
                   return typemin(T)
               end
           end
       end
f_long (generic function with 1 method)

julia> function f_short(posts)
           Ts = (UInt8, UInt16, UInt32, UInt64)
           i = findfirst(T -> length(posts) < typemax(T), Ts)
           return typemin(Ts[i])
       end
f_short (generic function with 1 method)

julia> @benchmark f_short(a) setup=begin a = rand(rand([10, 100, 5000, 50000, 10^6])) end
BenchmarkTools.Trial: 10000 samples with 962 evaluations.
 Range (min … max):  85.252 ns … 138.394 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     87.793 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   90.372 ns ±   6.200 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  ▁▅▇▇▅▃▄█▇▃                                    ▁▁▄▅▆▅▃▂       ▂
  ██████████▅▆▆▇█▇█▇█▇▆▇█▇█▇▆▇██▆▆▄▅▃▅▄▄▅▃▄▅▄▃▁▅█████████▆▅▅▃▅ █
  85.3 ns       Histogram: log(frequency) by time       105 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark f_long(a) setup=begin a = rand(rand([10, 100, 5000, 50000, 10^6])) end
BenchmarkTools.Trial: 10000 samples with 585 evaluations.
 Range (min … max):  202.494 ns …   1.774 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     325.138 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   316.423 ns ± 111.901 ns  ┊ GC (mean ± σ):  0.07% ± 1.06%

  █▃▂▂                 ▇▆▃▁▁▂                      ▆▄▂          ▂
  ████▇▆▄▄▄▄▄▃▄▃▂▆▄▅▅▂▃██████▇▇▇▅▆▆▄▂▅▂▄▅▅▅▅▃▄▃▄▄▃▄████▇▇▇█▇▅▅▅ █
  202 ns        Histogram: log(frequency) by time        547 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```

this is very funny

Reply:
I know that! I microbenchmarked that too before suggesting it

Reply:
yeah computers are hard LOL, let's just keep the loop anyway

Reply:
What's the codegen difference?

Reply:
there are somehow more boxing for long:
```julia
julia> @code_warntype f_long([1,2,3])
MethodInstance for f_long(::Vector{Int64})
  from f_long(posts) @ Main REPL[10]:1
Arguments
  #self#::Core.Const(f_long)
  posts::Vector{Int64}
Locals
  @_3::Union{Nothing, Tuple{DataType, Int64}}
  T::DataType
Body::Any
1 ─ %1  = Core.tuple(Main.UInt8, Main.UInt16, Main.UInt32, Main.UInt64)::Core.Const((UInt8, UInt16, UInt32, UInt64))
│         (@_3 = Base.iterate(%1))
│   %3  = (@_3::Core.Const((UInt8, 2)) === nothing)::Core.Const(false)
│   %4  = Base.not_int(%3)::Core.Const(true)
└──       goto #6 if not %4
2 ┄ %6  = @_3::Tuple{DataType, Int64}
│         (T = Core.getfield(%6, 1))
│   %8  = Core.getfield(%6, 2)::Int64
│   %9  = Main.length(posts)::Int64
│   %10 = Main.typemax(T)::Any
│   %11 = (%9 < %10)::Any
└──       goto #4 if not %11
3 ─ %13 = Main.typemin(T)::Any
└──       return %13
4 ─       (@_3 = Base.iterate(%1, %8))
│   %16 = (@_3 === nothing)::Bool
│   %17 = Base.not_int(%16)::Bool
└──       goto #6 if not %17
5 ─       goto #2
6 ┄       return nothing

julia> @code_warntype f_short([1,2,3])
MethodInstance for f_short(::Vector{Int64})
  from f_short(posts) @ Main REPL[11]:1
Arguments
  #self#::Core.Const(f_short)
  posts::Vector{Int64}
Locals
  #7::var"#7#8"{Vector{Int64}}
  i::Union{Nothing, Int64}
  Ts::NTuple{4, DataType}
Body::Any
1 ─      (Ts = Core.tuple(Main.UInt8, Main.UInt16, Main.UInt32, Main.UInt64))
│   %2 = Main.:(var"#7#8")::Core.Const(var"#7#8")
│   %3 = Core.typeof(posts)::Core.Const(Vector{Int64})
│   %4 = Core.apply_type(%2, %3)::Core.Const(var"#7#8"{Vector{Int64}})
│        (#7 = %new(%4, posts))
│   %6 = #7::var"#7#8"{Vector{Int64}}
│        (i = Main.findfirst(%6, Ts::Core.Const((UInt8, UInt16, UInt32, UInt64))))
│   %8 = Base.getindex(Ts::Core.Const((UInt8, UInt16, UInt32, UInt64)), i)::DataType
│   %9 = Main.typemin(%8)::Any
└──      return %9
```